spring-graphql/src/main/java/org/springframework/graphql/data/method/annotation/support/AuthenticationPrincipalArgumentResolver.java

```diff
@@ -0,0 +1,145 @@
+import org.springframework.util.StringUtils;
+import reactor.core.publisher.Mono;
+
+import java.lang.annotation.Annotation;
+
+/**
+ * Resolver to obtain {@link Authentication#getPrincipal()} from Spring Security context via
+ * {@link SecurityContext#getAuthentication()} for parameters annotated with {@link AuthenticationPrincipal}.
+ *
+ * <p>The resolver checks both ThreadLocal context via {@link SecurityContextHolder}
+ * for Spring MVC applications, and {@link ReactiveSecurityContextHolder} for
+ * Spring WebFlux applications.
+ *
+ * @author Rob Winch
+ * @since 1.0.0
+ */
+public class AuthenticationPrincipalArgumentResolver implements HandlerMethodArgumentResolver {
+
+	private ExpressionParser parser = new SpelExpressionParser();
+
+	private BeanResolver beanResolver;
+
+	/**
+	 * Sets the {@link BeanResolver} to be used on the expressions
+	 * @param beanResolver the {@link BeanResolver} to use
+	 */
+	public void setBeanResolver(BeanResolver beanResolver) {
+		this.beanResolver = beanResolver;
+	}
+
+	@Override
+	public boolean supportsParameter(MethodParameter parameter) {
+		return findMethodAnnotation(AuthenticationPrincipal.class, parameter) != null;
+	}
+
+	@Override
+	public Object resolveArgument(MethodParameter parameter, DataFetchingEnvironment environment) throws Exception {
+		return getCurrentAuthentication().map(Authentication::getPrincipal)
+				.map((principal) -> resolvePrincipal(parameter, principal))
+				.transform((argument) -> {
+					Class<?> parameterType = parameter.getParameterType();
+					boolean isParameterPublisher = Publisher.class.isAssignableFrom(parameterType);
+					return isParameterPublisher ? Mono.just(argument) : argument;
+				});
+	}
+
+	private Mono<Authentication> getCurrentAuthentication() {
+		SecurityContext securityContext = SecurityContextHolder.getContext();
+		return Mono.justOrEmpty(securityContext.getAuthentication())
+				.switchIfEmpty(ReactiveSecurityContextHolder.getContext().map(SecurityContext::getAuthentication));
+	}
+
+	private Object resolvePrincipal(MethodParameter parameter, Object principal) {
+		AuthenticationPrincipal annotation = findMethodAnnotation(AuthenticationPrincipal.class, parameter);
+		String expressionToParse = annotation.expression();
+		if (StringUtils.hasLength(expressionToParse)) {
+			StandardEvaluationContext context = new StandardEvaluationContext();
+			context.setRootObject(principal);
+			context.setVariable("this", principal);
+			context.setBeanResolver(this.beanResolver);
+			Expression expression = this.parser.parseExpression(expressionToParse);
+			principal = expression.getValue(context);
+		}
+		if (isInvalidType(parameter, principal)) {
+			if (annotation.errorOnInvalidType()) {
+				throw new ClassCastException(principal + " is not assignable to " + parameter.getParameterType());
+			}
+			return null;
+		}
+		return principal;
+	}
+
+	private boolean isInvalidType(MethodParameter parameter, Object principal) {
```

Comment:
`principal` should be marked `@Nullable`.

Reply:
This is now done.